sstables/sstables.hh

```diff
@@ -221,6 +222,12 @@ public:
     // will move it into a quarantine_dir subdirectory of its current directory.
     future<> move_to_quarantine(delayed_commit_changes* delay = nullptr);
 
+    // Move the sstable between states
+    //
+    // Known states are normal, staging, upload and quarantine.
+    // It's up to the storage driver how to implement this.
+    future<> change_state(sstring to, generation_type new_generation, delayed_commit_changes* delay = nullptr);
+
```

Comment:
we rely on stable generations in the sstable set reader. we don't have users changing sstable generation from X to Y. so we can kill new_generation param. we don't need to do it in this patch though.

Reply:
> we don't have users changing sstable generation from X to Y

Not quite

```c++
distributed_loader::make_sstables_available() {
    ...
        auto gen = table.calculate_generation_for_new_table();
        co_await sst->change_state(state, gen);
    ...
```

Previous version mitigated this by introducing specific methods to change states, in particular there was `move_from_upload` method that accepted target generation parameter

Reply:
you're right. that's only needed when moving from upload to avoid generation clash. less of a problem with UUID. although user could mistankely upload the same sstable more than once, and that should still work.

as the change state interface is now generic in the latest version, it becomes hard to put contraints on specific state switches. if we're moving a sstable in-use (e.g. in base directory), changing generation should be disallowed.